pkg/controller/syncmachineset/gcpactuator.go

```diff
@@ -0,0 +1,126 @@
+		return nil, err
+	}
+	actuator := &GCPActuator{
+		client: gcpClient,
+		logger: logger,
+		region: region,
+	}
+	return actuator, nil
+}
+
+// GenerateMachineSets satisfies the Actuator interface and will take a clusterDeployment and return a list of MachineSets
+// to sync to the remote cluster.
+func (a *GCPActuator) GenerateMachineSets(cd *hivev1.ClusterDeployment, ic *installtypes.InstallConfig, logger log.FieldLogger) ([]*machineapi.MachineSet, error) {
+	machineSets := []*machineapi.MachineSet{}
+
+	// get image ID for the generated machine sets
+	imageID, err := a.getImageID(cd, logger)
+	if err != nil {
+		a.logger.WithError(err).Warn("failed to find image ID for the machine sets")
+		return nil, err
+	}
+
+	// create the machinesets
+	for _, computePool := range ic.Compute {
+		if computePool.Platform.GCP == nil {
+			computePool.Platform.GCP = &installtypesgcp.MachinePool{}
+		}
+		if len(computePool.Platform.GCP.Zones) == 0 {
+			zones, err := getGCPZones(a.client, a.region)
+			if err != nil {
+				return nil, err
+			}
+
+			if len(zones) == 0 {
+				return nil, fmt.Errorf("fetched 0 zones from cloud provider")
+			}
+			computePool.Platform.GCP.Zones = zones
+		}
+
+		installerMachineSets, err := installgcp.MachineSets(cd.Status.InfraID, ic, &computePool, imageID, computePool.Name, "worker-user-data")
+		if err != nil {
+			logger.WithError(err).Error("failed to generate machinesets")
+			return nil, err
+		}
+
+		for _, ms := range installerMachineSets {
+			machineSets = append(machineSets, ms)
+		}
+	}
+
+	return machineSets, nil
+}
+
+func getGCPZones(client gcpclient.Client, region string) ([]string, error) {
+	zones := []string{}
+
+	// Filter to regions matching '.*<region>.*'
+	zoneFilter := fmt.Sprintf("region eq '.*%s.*'", region)
+	zoneList, err := client.ListComputeZones(gcpclient.ListComputeZonesOptions{Filter: zoneFilter})
+	if err != nil {
+		return zones, err
+	}
+
+	for _, zone := range zoneList.Items {
+		zones = append(zones, zone.Name)
+	}
+
+	return zones, nil
+}
+
+func (a *GCPActuator) getImageID(cd *hivev1.ClusterDeployment, logger log.FieldLogger) (string, error) {
+	infra := cd.Status.InfraID
+
+	// find names of the form '<infra>-.*'
+	filter := fmt.Sprintf("name eq \"%s-.*\"", infra)
+	result, err := a.client.ListComputeImages(gcpclient.ListComputeImagesOptions{Filter: filter})
+	if err != nil {
+		logger.WithError(err).Warnf("failed to find a GCP image starting with name: %s", infra)
+		return "", err
+	}
+	switch len(result.Items) {
+	case 0:
+		msg := fmt.Sprintf("found 0 results searching for GCP image starting with name: %s", infra)
+		logger.Warnf(msg)
+		return "", fmt.Errorf(msg)
+	case 1:
+		logger.Debugf("using image with name %s for machine sets", result.Items[0].Name)
+		return result.Items[0].Name, nil
+	default:
+		msg := fmt.Sprintf("unexpected number of results when looking for GCP image with name starting with %s", infra)
```

Comment:
an excellent question. what sequence of events would leave us with with multiple images when running the openshift-installer?

Reply:
I don't know. I don't think that the installer would generate more than just the one. But I am picturing a scenario where there are, say, RHEL nodes running from a different image whose name also starts with the infra ID prefix.

Reply:
i could change the regex to be more like what the installer is using https://github.com/openshift/installer/blob/master/pkg/asset/machines/gcp/machines.go#L89
wdyt?

Reply:
Might as well. Although at that point is there any need to even query GCP since we know the name already?

Reply:
i guess all we get is an error bubbled up that the image is missing out in the cloud. i'd rather not write out a known bad machineset in case the cluster happens to be limping along in whatever config it is currently in.